src/app/SubscriptionResumptionSessionEstablisher.cpp

```diff
@@ -0,0 +1,112 @@
+    mOnConnectedCallback(HandleDeviceConnected, this), mOnConnectionFailureCallback(HandleDeviceConnectionFailure, this)
+{}
+
+CHIP_ERROR
+SubscriptionResumptionSessionEstablisher::ResumeSubscription(
+    CASESessionManager & caseSessionManager, const SubscriptionResumptionStorage::SubscriptionInfo & subscriptionInfo)
+{
+    mSubscriptionInfo.mNodeId         = subscriptionInfo.mNodeId;
+    mSubscriptionInfo.mFabricIndex    = subscriptionInfo.mFabricIndex;
+    mSubscriptionInfo.mSubscriptionId = subscriptionInfo.mSubscriptionId;
+    mSubscriptionInfo.mMinInterval    = subscriptionInfo.mMinInterval;
+    mSubscriptionInfo.mMaxInterval    = subscriptionInfo.mMaxInterval;
+    mSubscriptionInfo.mFabricFiltered = subscriptionInfo.mFabricFiltered;
+    // Copy the Attribute Paths and Event Paths
+    if (subscriptionInfo.mAttributePaths.AllocatedSize() > 0)
+    {
+        mSubscriptionInfo.mAttributePaths.Alloc(subscriptionInfo.mAttributePaths.AllocatedSize());
+        if (!mSubscriptionInfo.mAttributePaths.Get())
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        for (size_t i = 0; i < mSubscriptionInfo.mAttributePaths.AllocatedSize(); ++i)
+        {
+            mSubscriptionInfo.mAttributePaths[i] = subscriptionInfo.mAttributePaths[i];
+        }
+    }
+    if (subscriptionInfo.mEventPaths.AllocatedSize() > 0)
+    {
+        mSubscriptionInfo.mEventPaths.Alloc(subscriptionInfo.mEventPaths.AllocatedSize());
+        if (!mSubscriptionInfo.mEventPaths.Get())
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        for (size_t i = 0; i < mSubscriptionInfo.mEventPaths.AllocatedSize(); ++i)
+        {
+            mSubscriptionInfo.mEventPaths[i] = subscriptionInfo.mEventPaths[i];
+        }
+    }
+
+    ScopedNodeId peerNode = ScopedNodeId(mSubscriptionInfo.mNodeId, mSubscriptionInfo.mFabricIndex);
+    caseSessionManager.FindOrEstablishSession(peerNode, &mOnConnectedCallback, &mOnConnectionFailureCallback);
+    return CHIP_NO_ERROR;
+}
+
+void SubscriptionResumptionSessionEstablisher::HandleDeviceConnected(void * context, Messaging::ExchangeManager & exchangeMgr,
+                                                                     const SessionHandle & sessionHandle)
+{
+    SubscriptionResumptionSessionEstablisher * _this = static_cast<SubscriptionResumptionSessionEstablisher *>(context);
+    SubscriptionResumptionStorage::SubscriptionInfo & subscriptionInfo = _this->mSubscriptionInfo;
+    InteractionModelEngine * imEngine                                  = InteractionModelEngine::GetInstance();
+    if (!imEngine->EnsureResourceForSubscription(subscriptionInfo.mFabricIndex, subscriptionInfo.mAttributePaths.AllocatedSize(),
+                                                 subscriptionInfo.mEventPaths.AllocatedSize()))
+    {
+        ChipLogProgress(InteractionModel, "no resource for subscription resumption");
+        delete _this;
```

Comment:
delete _this is repeated on all return paths. This is error prone - make the deletion RAII (code should look like `DeleteOnReturn deleter(_this);` for some `DeleteOnReturn` class/templated class that you create in this cpp file. it should just have constructor and a destructor that calls delete.

Reply:
Done